ipam/ipam.go

```diff
@@ -0,0 +1,250 @@
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "verifying that subnet allocations don't overlap")
+
+		for _, cn := range clusters {
+			awsConfig, err := c.hostFramework.AWSCluster(cn)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			c.logger.LogCtx(ctx, "level", "debug", "message", "verify that there are no duplicate subnet allocations")
+			subnet := awsConfig.Status.Cluster.Network.CIDR
+			otherCluster, exists := allocatedSubnets[subnet]
+			if exists {
+				return microerror.Maskf(alreadyExistsError, "subnet %s already exists for %s", subnet, otherCluster)
+			}
+
+			// Verify that allocated subnets don't overlap.
+			for k, _ := range allocatedSubnets {
+				err := verifyNoOverlap(subnet, k)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+
+			allocatedSubnets[subnet] = cn
+		}
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("terminating guest cluster %s and immediately creating new guest cluster %s", clusterTwo, clusterFour))
+
+		err = c.provider.DeleteCluster(clusterTwo)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		err = c.provider.CreateCluster(clusterFour)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for guest cluster %s to shutdown", clusterTwo))
+
+		guest := guestFrameworks[clusterTwo]
+		err = guest.WaitForAPIDown()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		delete(guestFrameworks, clusterTwo)
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("guest cluster %s down", clusterTwo))
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for guest cluster %s to come up", clusterFour))
+
+		cfg := framework.GuestConfig{
+			Logger: c.logger,
+
+			ClusterID:    clusterFour,
+			CommonDomain: c.commonDomain,
+		}
+		guestFramework, err := framework.NewGuest(cfg)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		guestFrameworks[clusterFour] = guestFramework
+		err = guestFramework.Setup()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("guest cluster %s up", clusterFour))
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "verify that clusterFour subnet doesn't overlap with other allocations")
+		awsConfig, err := c.hostFramework.AWSCluster(clusterFour)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		subnet := awsConfig.Status.Cluster.Network.CIDR
+		otherCluster, exists := allocatedSubnets[subnet]
+		if exists {
+			return microerror.Maskf(alreadyExistsError, "subnet %s already exists for %s", subnet, otherCluster)
+		}
+
+		for k, _ := range allocatedSubnets {
+			err := verifyNoOverlap(subnet, k)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	return err
+}
+
+func verifyNoOverlap(subnet1, subnet2 string) error {
```

Comment:
Does it make sense to unit test it?

Reply:
Unsure. `ipam.Contains()` has tests and this function is pretty simple. I can do it if you see it useful here.